examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,1673 @@
+                student_index = torch.randint(0, student_distillation_steps, (bsz,), device=latents.device).long()
+                student_timesteps = student_timestep_schedule[student_index]
+                teacher_timesteps = torch.randint(
+                    0, noise_scheduler.config.num_train_timesteps, (bsz,), device=latents.device
+                ).long()
+
+                # 3. Sample noise and add it to the latents according to the noise magnitude at each student timestep
+                # (that is, run the forward process on the student model)
+                student_noise = torch.randn_like(latents)
+                noisy_student_input = noise_scheduler.add_noise(latents, student_noise, student_timesteps)
+
+                # 4. Prepare prompt embeds and unet_added_conditions
+                prompt_embeds = encoded_text.pop("prompt_embeds")
+
+                # 5. Get the student model predicted original sample `student_x_0`.
+                student_noise_pred = unet(
+                    noisy_student_input,
+                    student_timesteps,
+                    encoder_hidden_states=prompt_embeds.float(),
+                ).sample
+                student_x_0 = denoiser(student_noise_pred, student_timesteps, noisy_student_input)
+
+                # 6. Sample noise and add it to the student's predicted original sample according to the noise
+                # magnitude at each teacher timestep (that is, run the forward process on the teacher model, but
+                # using `student_x_0` instead of latents sampled from the prior).
+                teacher_noise = torch.randn_like(student_x_0)
+                noisy_teacher_input = noise_scheduler.add_noise(student_x_0, teacher_noise, teacher_timesteps)
+
+                # 7. Get teacher model predicted original sample `teacher_x_0`.
+                with torch.no_grad():
+                    with torch.autocast("cuda"):
+                        teacher_noise_pred = teacher_unet(
+                            noisy_teacher_input.detach(),
+                            teacher_timesteps,
+                            encoder_hidden_states=prompt_embeds.float(),
+                        ).sample
+                        teacher_x_0 = denoiser(teacher_noise_pred, teacher_timesteps, noisy_teacher_input)
+
+                ############################
+                # 8. Discriminator Loss
+                ############################
+                discriminator_optimizer.zero_grad(set_to_none=True)
+
+                # 1. Decode real and fake (generated) latents back to pixel space.
+                # NOTE: the paper doesn't mention this explicitly AFAIK but I think this makes sense since the
+                # pretrained feature network for the discriminator operates in pixel space rather than latent space.
+                real_image = vae.decode(latents).sample
+                student_gen_image = vae.decode(student_x_0).sample
```

Comment:
We should unscale the latents before decoding them.

Also, we should add an option for decoding `n` batches at a time. As decoding the whole batch might run into OOM.

Reply:
Added batched VAE decoding and made the VAE encoding/decoding batch size configurable via the `vae_encode_batch_size` argument.